File: 🟡 Medium/2415 - Reverse Odd Levels of Binary Tree/_2415ReverseOddLevelsOfBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var reverseOddLevels = function (root) {
    if (!root) return null;

    dfs(root.left, root.right, 1);

    return root;
};

function dfs(node1, node2, level) {
    if (node1 === null || node2 === null) {
        return;
    }

    if (level % 2 !== 0) {
        const temp = node1.val;
        node1.val = node2.val;
        node2.val = temp;
    }

    dfs(node1.left, node2.right, level + 1);
    dfs(node1.right, node2.left, level + 1);
}

File: 🟡 Medium/2415 - Reverse Odd Levels of Binary Tree/_2415ReverseOddLevelsOfBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def reverseOddLevels(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        q = deque([root])

        level = False

        while q:

            if level:
                l, r = 0, len(q) - 1
                while l < r:

                    q[l].val, q[r].val = q[r].val, q[l].val
                    l += 1
                    r -= 1

            for _ in range(len(q)):
                node = q.popleft()
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)

            level = not level

        return root